File: ios/SiezaGym/Services/GymStore.swift
import Foundation
import Observation
import os

private let log = Logger(subsystem: "com.siezagym.app", category: "store")

/// Estado compartido de la app: una sola carga de Firestore alimenta las cinco
/// pantallas. Si cada tab consultara por su cuenta pagariamos las mismas
/// lecturas cinco veces y podrian mostrar numeros distintos entre si.
@Observable
final class GymStore {
    let uid: String

    private(set) var profile: UserProfile?
    private(set) var routines: [Routine] = []
    private(set) var sessions: [WorkoutSession] = []
    private(set) var catalog: [String: Exercise] = [:]
    private(set) var isLoading = false
    private(set) var loadError: String?
    /// Nil hasta la primera carga: sirve para no mostrar "no hay nada" mientras carga.
    private(set) var lastLoadedAt: Date?

    /// Rutina en curso mantenida en stand-by
    var activeWorkout: WorkoutDraft?

    let healthKit = HealthKitService()

    private let repository = GymRepository()

    init(uid: String) {
        self.uid = uid
    }

    var hasLoaded: Bool { lastLoadedAt != nil }

    func load() async {
        guard !isLoading else { return }
        isLoading = true
        defer { isLoading = false }

        do {
            // Las cuatro consultas son independientes: van en paralelo.
            async let catalog = repository.exercises(uid: uid)
            async let profile = repository.profile(uid: uid)
            async let routines = repository.routines(uid: uid)
            async let sessions = repository.sessions(uid: uid)

            self.catalog = try await catalog
            self.profile = try await profile
            self.routines = try await routines
            self.sessions = try await sessions
            loadError = nil
            lastLoadedAt = Date()
            publicarWidget()
        } catch {
            log.error("carga fallo: \(error.localizedDescription, privacy: .public)")
            loadError = "No pudimos traer tus datos. Deslizá para reintentar."
        }
    }

    /// El widget corre en otro proceso y no llega a Firestore: la app le deja
    /// el resumen escrito cada vez que trae datos.
    private func publicarWidget() {
        WidgetBridge.publicar(
            WidgetSnapshotBuilder.build(
                themeID: ThemeStore.temaGuardado,
                sessions: sessions,
                routine: featuredRoutine,
                catalog: catalog
            )
        )
    }

    func exercise(_ id: String) -> Exercise? { catalog[id] }

    func name(of exerciseID: String) -> String {
        catalog[exerciseID]?.nameEs ?? exerciseID
    }

    // MARK: - Derivados de la Home

    var trainedDayKeys: Set<String> { HomeMetrics.trainedDayKeys(sessions) }

    var streak: Int { TrainingCalendar.streak(trainedDayKeys: trainedDayKeys) }

    /// La rutina que la Home propone: la ultima usada que se muestre en Home.
    var featuredRoutine: Routine? {
        routines.first { $0.showOnHome } ?? routines.first
    }

    var weekSessions: [WorkoutSession] { HomeMetrics.sessionsInLastDays(sessions, days: 7) }

    var weeklyVolumeKg: Double { weekSessions.reduce(0) { $0 + $1.totalVolumeKg } }

    var muscleVolume: HomeMetrics.MuscleVolume {
        HomeMetrics.volumeByMuscleGroup(sessions, catalog: catalog)
    }

    var pushPull: HomeMetrics.PushPull { HomeMetrics.pushPullBalance(sessions, catalog: catalog) }

    var completion: HomeMetrics.Completion { HomeMetrics.setCompletionRate(sessions) }

    var weekdayVolume: [HomeMetrics.WeekdayVolume] { HomeMetrics.volumeByWeekday(sessions) }

    var intensity: HomeMetrics.Intensity { HomeMetrics.relativeIntensity(sessions) }

    var volumeTrend: HomeMetrics.VolumeTrend { HomeMetrics.volumePerSession(sessions) }

    var zones: HomeMetrics.Zones { HomeMetrics.intensityZones(sessions) }

    var calories: HomeMetrics.CalorieGoal {
        HomeMetrics.weeklyCalories(
            weekSessions,
            bodyWeightKg: profile?.bodyWeightKg,
            goal: profile?.weeklyCalorieGoalKcal
        )
    }

    // MARK: - Escrituras

    func saveSession(routine: Routine?, startedAt: Date, exercises: [LoggedExercise]) async throws {
        try await repository.saveSession(
            uid: uid,
            routine: routine,
            startedAt: startedAt,
            exercises: exercises
        )
        await load()
    }

    func deleteSession(_ session: WorkoutSession) async throws {
        guard session.userID == uid else { return }
        try await repository.deleteSession(uid: uid, sessionID: session.id)
        sessions.removeAll { $0.id == session.id }
    }

    /// Crea una rutina y recarga, para que aparezca en la lista sin salir y
    /// volver a entrar.
    func createRoutine(name: String, note: String, exercises: [RoutineDraftExercise]) async throws {
        try await repository.createRoutine(uid: uid, name: name, note: note, exercises: exercises)
        await load()
    }

    /// Guarda los cambios de una rutina y recarga, para que el detalle muestre
    /// lo editado sin salir y volver a entrar.
    func updateRoutine(_ routine: Routine, name: String, note: String, exercises: [RoutineDraftExercise]) async throws {
        // Una rutina del coach vive en `assignments` y no es del alumno.
        guard !routine.isAssigned else { throw RoutineEditError.esDelCoach }
        try await repository.updateRoutine(
            routineID: routine.id,
            name: name,
            note: note,
            exercises: exercises
        )
        await load()
    }

    /// Crea un ejercicio propio y lo suma al catálogo en memoria, para que
    /// aparezca en el selector sin recargar todo desde Firestore.
    @discardableResult
    func createCustomExercise(_ draft: CustomExerciseDraft) async throws -> Exercise {
        let ejercicio = try await repository.createCustomExercise(uid: uid, draft: draft)
        catalog[ejercicio.id] = ejercicio
        return ejercicio
    }

    func updateProfile(_ fields: [String: Any]) async {
        do {
            try await repository.updateProfile(uid: uid, fields: fields)
            profile = try await repository.profile(uid: uid)
        } catch {
            log.error("perfil no se guardo: \(error.localizedDescription, privacy: .public)")
            loadError = "No se pudo guardar el perfil."
        }
    }
}

nonisolated enum RoutineEditError: LocalizedError {
    case esDelCoach

    var errorDescription: String? {
        "Las rutinas del coach se editan desde su panel."
    }
}

File: android/app/src/main/kotlin/com/siezagym/app/Services/GymStore.kt
package com.siezagym.app.Services

import com.siezagym.app.Domain.HomeMetrics
import com.siezagym.app.Domain.RoutineDraftExercise
import com.siezagym.app.Domain.TrainingCalendar
import com.siezagym.app.Models.Exercise
import com.siezagym.app.Models.LoggedExercise
import com.siezagym.app.Models.Routine
import com.siezagym.app.Models.UserProfile
import com.siezagym.app.Models.WorkoutSession
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.ZonedDateTime

/** Estado compartido de la app: una sola carga de Firestore alimenta las cinco
 *  pantallas. Si cada tab consultara por su cuenta pagaríamos las mismas
 *  lecturas cinco veces y podrían mostrar números distintos entre sí. */
class GymStore(val uid: String) {
    private val repository = GymRepository()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private val _profile = MutableStateFlow<UserProfile?>(null)
    private val _routines = MutableStateFlow<List<Routine>>(emptyList())
    private val _sessions = MutableStateFlow<List<WorkoutSession>>(emptyList())
    private val _catalog = MutableStateFlow<Map<String, Exercise>>(emptyMap())
    private val _isLoading = MutableStateFlow(false)
    private val _loadError = MutableStateFlow<String?>(null)
    /** Null hasta la primera carga: sirve para no mostrar "no hay nada" mientras carga. */
    private val _lastLoadedAt = MutableStateFlow<ZonedDateTime?>(null)

    val profile: StateFlow<UserProfile?> = _profile.asStateFlow()
    val routines: StateFlow<List<Routine>> = _routines.asStateFlow()
    val sessions: StateFlow<List<WorkoutSession>> = _sessions.asStateFlow()
    val catalog: StateFlow<Map<String, Exercise>> = _catalog.asStateFlow()
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    val loadError: StateFlow<String?> = _loadError.asStateFlow()
    val lastLoadedAt: StateFlow<ZonedDateTime?> = _lastLoadedAt.asStateFlow()

    val hasLoaded: Boolean
        get() = _lastLoadedAt.value != null

    fun load() {
        scope.launch { refresh() }
    }

    private suspend fun refresh() {
        if (_isLoading.value) return
        _isLoading.value = true
        try {
            // Las cuatro consultas son independientes: van en paralelo.
            coroutineScope {
                val catalog = async { repository.exercises(uid) }
                val profile = async { repository.profile(uid) }
                val routines = async { repository.routines(uid) }
                val sessions = async { repository.sessions(uid) }

                _catalog.value = catalog.await()
                _profile.value = profile.await()
                _routines.value = routines.await()
                _sessions.value = sessions.await()
            }
            _loadError.value = null
            _lastLoadedAt.value = ZonedDateTime.now(TrainingCalendar.zone)
        } catch (e: Exception) {
            _loadError.value = "No pudimos traer tus datos. Volvé a intentar."
        } finally {
            _isLoading.value = false
        }
    }

    fun exercise(id: String): Exercise? = _catalog.value[id]

    fun name(of exerciseID: String): String = _catalog.value[exerciseID]?.nameEs ?: exerciseID

    fun rutina(id: String): Routine? = _routines.value.firstOrNull { it.id == id }

    // MARK: - Derivados de la Home

    val trainedDayKeys: Set<String>
        get() = HomeMetrics.trainedDayKeys(_sessions.value)

    val streak: Int
        get() = TrainingCalendar.streak(trainedDayKeys)

    /** La rutina que la Home propone: la última usada que se muestre en Home. */
    val featuredRoutine: Routine?
        get() = _routines.value.firstOrNull { it.showOnHome } ?: _routines.value.firstOrNull()

    val weekSessions: List<WorkoutSession>
        get() = HomeMetrics.sessionsInLastDays(_sessions.value, days = 7)

    val weeklyVolumeKg: Double
        get() = weekSessions.sumOf { it.totalVolumeKg }

    val muscleVolume: HomeMetrics.MuscleVolume
        get() = HomeMetrics.volumeByMuscleGroup(_sessions.value, _catalog.value)

    val pushPull: HomeMetrics.PushPull
        get() = HomeMetrics.pushPullBalance(_sessions.value, _catalog.value)

    val completion: HomeMetrics.Completion
        get() = HomeMetrics.setCompletionRate(_sessions.value)

    val weekdayVolume: List<HomeMetrics.WeekdayVolume>
        get() = HomeMetrics.volumeByWeekday(_sessions.value)

    val intensity: HomeMetrics.Intensity
        get() = HomeMetrics.relativeIntensity(_sessions.value)

    val volumeTrend: HomeMetrics.VolumeTrend
        get() = HomeMetrics.volumePerSession(_sessions.value)

    val zones: HomeMetrics.Zones
        get() = HomeMetrics.intensityZones(_sessions.value)

    val calories: HomeMetrics.CalorieGoal
        get() = HomeMetrics.weeklyCalories(
            weekSessions,
            bodyWeightKg = _profile.value?.bodyWeightKg,
            goal = _profile.value?.weeklyCalorieGoalKcal,
        )

    // MARK: - Escrituras

    suspend fun saveSession(routine: Routine?, startedAt: Instant, exercises: List<LoggedExercise>) {
        repository.saveSession(uid = uid, routine = routine, startedAt = startedAt, exercises = exercises)
        refresh()
    }

    suspend fun createRoutine(name: String, note: String, exercises: List<RoutineDraftExercise>) {
        repository.createRoutine(uid = uid, name = name, note = note, exercises = exercises)
        refresh()
    }

    suspend fun updateProfile(fields: Map<String, Any?>) {
        try {
            repository.updateProfile(uid, fields)
            _profile.value = repository.profile(uid)
        } catch (e: Exception) {
            _loadError.value = "No se pudo guardar el perfil."
        }
    }
}